### ISSUE TITLE
Why alignment loss will degrades model performance in City2Foggy?

### ISSUE THREAD

**Issue by 123456184** [TRAILab/DINO_Teacher#11]
Using alignment losses in the Cityscapes to Foggy Cityscapes dataset makes it less effective. Whether it is MLP, MLP3, or Attention, it will make the performance worse, but the best performance will be achieved without alignment loss

**Comment by zqf30** [comment]
Hi @123456184 , this is an interesting question. I wonder if it is casused by the similarity loss. Maybe L1, L2 would be effective?

**Comment by fnsfd** [comment]
You can set MIN_SIZE_TRAIN: (550,) to MIN_SIZE_TRAIN: (600,). Maybe it would be effective?

**Comment by zqf30** [comment]
I got the same result with @123456184 . And I also have done the ablation study on cityscapes 2 foggy cityscapes. (1) Baseline. Train on VGG-based Faster RCNN. The mAP50 in foggy cityscapes (1.5k validation) reach to 46.0. (2) Train on VGG-based Faster RCNN, with the proposed alignment, and w/o the pseudo label training. the mAP drops to 40.0. (3)Train on VGG-based Faster RCNN, w/0 the proposed alignment, and with the pseudo label training. The mAP reach to 54.7. It seems that the performances of baseline and the proposed method are not correct.

ps: I also need to note that in the implementation, the vgg pretrained-ckpt is not loaded due to the wrong coding when loading ckpt.   I rectify this error, and have done the above experiment.

**Comment by zqf30** [comment]
> You can set MIN_SIZE_TRAIN: (550,) to MIN_SIZE_TRAIN: (600,). Maybe it would be effective?

Hi fnsfd. Did you reproduce the experiment result successfully using this setting?

**Comment by Kaliyah67** [comment]
> I got the same result with [@123456184](https://github.com/123456184) . And I also have done the ablation study on cityscapes 2 foggy cityscapes. (1) Baseline. Train on VGG-based Faster RCNN. The mAP50 in foggy cityscapes (1.5k validation) reach to 46.0. (2) Train on VGG-based Faster RCNN, with the proposed alignment, and w/o the pseudo label training. the mAP drops to 40.0. (3)Train on VGG-based Faster RCNN, w/0 the proposed alignment, and with the pseudo label training. The mAP reach to 54.7. It seems that the performances of baseline and the proposed method are not correct.
> 
> ps: I also need to note that in the implementation, the vgg pretrained-ckpt is not loaded due to the wrong coding when loading ckpt. I rectify this error, and have done the above experiment.

Hi zqf30,

Thank you for sharing your ablation study results! That's really helpful.

I am currently trying to reproduce the same experiment (Cityscapes to Foggy Cityscapes with the pseudo label training and without the proposed alignment), but my mAP only reached 51.9.

Could you please let me know which part of the code you modified to ensure the VGG pretrained-ckpt loaded correctly?

**Comment by zqf30** [comment]
@Kaliyah67 Hi, sorry for the late response. I add the loading operation in line96-98, in DINO_Teacher-main/adapteacher/modeling/meta_arch/vgg.py:
```
# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
import torch.nn as nn
import copy
import torch
from typing import Union, List, Dict, Any, cast
from detectron2.modeling.backbone import (
    ResNet,
    Backbone,
    build_resnet_backbone,
    BACKBONE_REGISTRY
)
from detectron2.modeling.backbone.fpn import FPN, LastLevelMaxPool, LastLevelP6P7



def make_layers(cfg: List[Union[str, int]], batch_norm: bool = False) -> nn.Sequential:
    layers: List[nn.Module] = []
    in_channels = 3
    for v in cfg:
        if v == 'M':
            layers += [nn.MaxPool2d(kernel_size=2, stride=2)]
        else:
            v = cast(int, v)
            conv2d = nn.Conv2d(in_channels, v, kernel_size=3, padding=1)
            if batch_norm:
                layers += [conv2d, nn.BatchNorm2d(v), nn.ReLU(inplace=True)]
            else:
                layers += [conv2d, nn.ReLU(inplace=True)]
            in_channels = v
    return nn.Sequential(*layers)

cfgs: Dict[str, List[Union[str, int]]] = {
    'vgg11': [64, 'M', 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'],
    'vgg13': [64, 64, 'M', 128, 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'],
    'vgg16': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 'M', 512, 512, 512, 'M', 512, 512, 512, 'M'],
    'vgg19': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 256, 'M', 512, 512, 512, 512, 'M', 512, 512, 512, 512, 'M'],
}


class vgg_backbone(Backbone):
    """
    Backbone (bottom-up) for FBNet.

    Hierarchy:
        trunk0:
            xif0_0
            xif0_1
            ...
        trunk1:
            xif1_0
            xif1_1
            ...
        ...

    Output features:
        The outputs from each "stage", i.e. trunkX.
    """

    def __init__(self, cfg):
        super().__init__()

        self.vgg = make_layers(cfgs['vgg16'],batch_norm=True)

        self._initialize_weights()
        # self.stage_names_index = {'vgg1':3, 'vgg2':8 , 'vgg3':15, 'vgg4':22, 'vgg5':29}
        _out_feature_channels = [64, 128, 256, 512, 512]
        _out_feature_strides = [2, 4, 8, 16, 32]
        # stages, shape_specs = build_fbnet(
        #     cfg,
        #     name="trunk",
        #     in_channels=cfg.MODEL.FBNET_V2.STEM_IN_CHANNELS
        # )

        # nn.Sequential(*list(self.vgg.features._modules.values())[:14])

        self.stages = [nn.Sequential(*list(self.vgg._modules.values())[0:7]),\
                    nn.Sequential(*list(self.vgg._modules.values())[7:14]),\
                    nn.Sequential(*list(self.vgg._modules.values())[14:24]),\
                    nn.Sequential(*list(self.vgg._modules.values())[24:34]),\
                    nn.Sequential(*list(self.vgg._modules.values())[34:]),]
        self._out_feature_channels = {}
        self._out_feature_strides = {}
        self._stage_names = []

        for i, stage in enumerate(self.stages):
            name = "vgg{}".format(i)
            self.add_module(name, stage)
            self._stage_names.append(name)
            self._out_feature_channels[name] = _out_feature_channels[i]
            self._out_feature_strides[name] = _out_feature_strides[i]

        self._out_features = self._stage_names

        del self.vgg
        
        if("vgg16" in cfg.MODEL.WEIGHTS):
            checkpoint = torch.load(cfg.MODEL.WEIGHTS, map_location='cuda')
            self.load_state_dict(checkpoint)
        

    def forward(self, x):
        features = {}
        for name, stage in zip(self._stage_names, self.stages):
            x = stage(x)
            # if name in self._out_features:
            #     outputs[name] = x
            features[name] = x
        # import pdb
        # pdb.set_trace()

        return features

    def _initialize_weights(self) -> None:
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.Linear):
                nn.init.normal_(m.weight, 0, 0.01)
                nn.init.constant_(m.bias, 0)


@BACKBONE_REGISTRY.register() #already register in baseline model
def build_vgg_backbone(cfg, _):
    return vgg_backbone(cfg)


@BACKBONE_REGISTRY.register() #already register in baseline model
def build_vgg_fpn_backbone(cfg, _):
    # backbone = FPN(
    #     bottom_up=build_vgg_backbone(cfg),
    #     in_features=cfg.MODEL.FPN.IN_FEATURES,
    #     out_channels=cfg.MODEL.FPN.OUT_CHANNELS,
    #     norm=cfg.MODEL.FPN.NORM,
    #     top_block=LastLevelMaxPool(),
    # )

    bottom_up = vgg_backbone(cfg)
    in_features = cfg.MODEL.FPN.IN_FEATURES
    out_channels = cfg.MODEL.FPN.OUT_CHANNELS
    backbone = FPN(
        bottom_up=bottom_up,
        in_features=in_features,
        out_channels=out_channels,
        norm=cfg.MODEL.FPN.NORM,
        top_block=LastLevelMaxPool(),
        # fuse_type=cfg.MODEL.FPN.FUSE_TYPE,
    )
    # return backbone

    return backbone
```

**Comment by slizm2025** [comment]
> [@Kaliyah67](https://github.com/Kaliyah67) Hi, sorry for the late response. I add the loading operation in line96-98, in DINO_Teacher-main/adapteacher/modeling/meta_arch/vgg.py:
> 
> ```
> # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
> import torch.nn as nn
> import copy
> import torch
> from typing import Union, List, Dict, Any, cast
> from detectron2.modeling.backbone import (
>     ResNet,
>     Backbone,
>     build_resnet_backbone,
>     BACKBONE_REGISTRY
> )
> from detectron2.modeling.backbone.fpn import FPN, LastLevelMaxPool, LastLevelP6P7
> 
> 
> 
> def make_layers(cfg: List[Union[str, int]], batch_norm: bool = False) -> nn.Sequential:
>     layers: List[nn.Module] = []
>     in_channels = 3
>     for v in cfg:
>         if v == 'M':
>             layers += [nn.MaxPool2d(kernel_size=2, stride=2)]
>         else:
>             v = cast(int, v)
>             conv2d = nn.Conv2d(in_channels, v, kernel_size=3, padding=1)
>             if batch_norm:
>                 layers += [conv2d, nn.BatchNorm2d(v), nn.ReLU(inplace=True)]
>             else:
>                 layers += [conv2d, nn.ReLU(inplace=True)]
>             in_channels = v
>     return nn.Sequential(*layers)
> 
> cfgs: Dict[str, List[Union[str, int]]] = {
>     'vgg11': [64, 'M', 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'],
>     'vgg13': [64, 64, 'M', 128, 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'],
>     'vgg16': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 'M', 512, 512, 512, 'M', 512, 512, 512, 'M'],
>     'vgg19': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 256, 'M', 512, 512, 512, 512, 'M', 512, 512, 512, 512, 'M'],
> }
> 
> 
> class vgg_backbone(Backbone):
>     """
>     Backbone (bottom-up) for FBNet.
> 
>     Hierarchy:
>         trunk0:
>             xif0_0
>             xif0_1
>             ...
>         trunk1:
>             xif1_0
>             xif1_1
>             ...
>         ...
> 
>     Output features:
>         The outputs from each "stage", i.e. trunkX.
>     """
> 
>     def __init__(self, cfg):
>         super().__init__()
> 
>         self.vgg = make_layers(cfgs['vgg16'],batch_norm=True)
> 
>         self._initialize_weights()
>         # self.stage_names_index = {'vgg1':3, 'vgg2':8 , 'vgg3':15, 'vgg4':22, 'vgg5':29}
>         _out_feature_channels = [64, 128, 256, 512, 512]
>         _out_feature_strides = [2, 4, 8, 16, 32]
>         # stages, shape_specs = build_fbnet(
>         #     cfg,
>         #     name="trunk",
>         #     in_channels=cfg.MODEL.FBNET_V2.STEM_IN_CHANNELS
>         # )
> 
>         # nn.Sequential(*list(self.vgg.features._modules.values())[:14])
> 
>         self.stages = [nn.Sequential(*list(self.vgg._modules.values())[0:7]),\
>                     nn.Sequential(*list(self.vgg._modules.values())[7:14]),\
>                     nn.Sequential(*list(self.vgg._modules.values())[14:24]),\
>                     nn.Sequential(*list(self.vgg._modules.values())[24:34]),\
>                     nn.Sequential(*list(self.vgg._modules.values())[34:]),]
>         self._out_feature_channels = {}
>         self._out_feature_strides = {}
>         self._stage_names = []
> 
>         for i, stage in enumerate(self.stages):
>             name = "vgg{}".format(i)
>             self.add_module(name, stage)
>             self._stage_names.append(name)
>             self._out_feature_channels[name] = _out_feature_channels[i]
>             self._out_feature_strides[name] = _out_feature_strides[i]
> 
>         self._out_features = self._stage_names
> 
>         del self.vgg
>         
>         if("vgg16" in cfg.MODEL.WEIGHTS):
>             checkpoint = torch.load(cfg.MODEL.WEIGHTS, map_location='cuda')
>             self.load_state_dict(checkpoint)
>         
> 
>     def forward(self, x):
>         features = {}
>         for name, stage in zip(self._stage_names, self.stages):
>             x = stage(x)
>             # if name in self._out_features:
>             #     outputs[name] = x
>             features[name] = x
>         # import pdb
>         # pdb.set_trace()
> 
>         return features
> 
>     def _initialize_weights(self) -> None:
>         for m in self.modules():
>             if isinstance(m, nn.Conv2d):
>                 nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
>                 if m.bias is not None:
>                     nn.init.constant_(m.bias, 0)
>             elif isinstance(m, nn.BatchNorm2d):
>                 nn.init.constant_(m.weight, 1)
>                 nn.init.constant_(m.bias, 0)
>             elif isinstance(m, nn.Linear):
>                 nn.init.normal_(m.weight, 0, 0.01)
>                 nn.init.constant_(m.bias, 0)
> 
> 
> @BACKBONE_REGISTRY.register() #already register in baseline model
> def build_vgg_backbone(cfg, _):
>     return vgg_backbone(cfg)
> 
> 
> @BACKBONE_REGISTRY.register() #already register in baseline model
> def build_vgg_fpn_backbone(cfg, _):
>     # backbone = FPN(
>     #     bottom_up=build_vgg_backbone(cfg),
>     #     in_features=cfg.MODEL.FPN.IN_FEATURES,
>     #     out_channels=cfg.MODEL.FPN.OUT_CHANNELS,
>     #     norm=cfg.MODEL.FPN.NORM,
>     #     top_block=LastLevelMaxPool(),
>     # )
> 
>     bottom_up = vgg_backbone(cfg)
>     in_features = cfg.MODEL.FPN.IN_FEATURES
>     out_channels = cfg.MODEL.FPN.OUT_CHANNELS
>     backbone = FPN(
>         bottom_up=bottom_up,
>         in_features=in_features,
>         out_channels=out_channels,
>         norm=cfg.MODEL.FPN.NORM,
>         top_block=LastLevelMaxPool(),
>         # fuse_type=cfg.MODEL.FPN.FUSE_TYPE,
>     )
>     # return backbone
> 
>     return backbone
> ```

Hi, @zqf30  and @Kaliyah67 . 
I am currently reproducing the Cityscapes to Foggy Cityscapes experiment and have encountered the same issue: when I enable the alignment module, the experimental performance actually degrades.

Could you please clarify if your reproduced results on this dataset are indeed 51.9 and 54.7, respectively, and whether these results were obtained strictly using the architecture proposed in the original paper? Furthermore, was the improvement between these two metrics achieved by @zqf30  utilizing the loading strategy mentioned above?

However, it appears that both of these results still fall short of the performance reported in the original paper (55.4).
